connectors/sources/dropbox_aiohttp.py

```diff
@@ -0,0 +1,659 @@
+        elif error_response == "invalid_client: Invalid client_id or client_secret":
+            raise InvalidClientCredentialException(
+                "Configured App Key or App Secret is invalid."
+            )
+        else:
+            raise Exception(
+                f"Error while generating an access token. Error: {error_response}"
+            )
+
+    @cached_property
+    def _get_session(self):
+        self._logger.debug("Generating aiohttp client session")
+        timeout = aiohttp.ClientTimeout(total=None)  # pyright: ignore
+
+        return aiohttp.ClientSession(timeout=timeout, raise_for_status=True)
+
+    async def close(self):
+        self._sleeps.cancel()
+        await self._get_session.close()
+        del self._get_session
+
+    def _get_request_headers(self, file_type, **kwargs):
+        kwargs = kwargs["kwargs"]
+        request_headers = {
+            "Authorization": f"Bearer {self.access_token}",
+        }
+        if file_type == FILE:
+            request_headers["Dropbox-API-Arg"] = f'{{"path": "{kwargs["path"]}"}}'
+        elif file_type == PAPER:
+            request_headers[
+                "Dropbox-API-Arg"
+            ] = f'{{"path": "{kwargs["path"]}", "export_format": "markdown"}}'
+        elif file_type == RECEIVED_FILE:
+            request_headers["Dropbox-API-Arg"] = f'{{"url": "{kwargs["url"]}"}}'
+        else:
+            request_headers["Content-Type"] = "application/json"
+        return request_headers
+
+    def _get_retry_after(self, retry, exception):
+        if retry > self.retry_count:
+            raise exception
+        self._logger.warning(
+            f"Retry count: {retry} out of {self.retry_count}. Exception: {exception}"
+        )
+        retry += 1
+        return retry, RETRY_INTERVAL**retry
+
+    async def _handle_client_errors(self, retry, exception):
+        retry, retry_seconds = self._get_retry_after(retry=retry, exception=exception)
+        match exception.status:
+            case 401:
+                await self._set_access_token()
+            case 409:
+                raise InvalidPathException(
+                    f"Configured Path: {self.path} is invalid or not found."
+                ) from exception
+            case 429:
+                response_headers = exception.headers
+                retry_seconds = int(
+                    response_headers.get("Retry-After", DEFAULT_RETRY_AFTER)
+                )
+                self._logger.warning(
+                    f"Rate limited by Dropbox. Retrying after {retry_seconds} seconds"
+                )
+        self._logger.debug(f"Will retry in {retry_seconds} seconds")
+        await self._sleeps.sleep(retry_seconds)
+        return retry
+
+    async def api_call(self, base_url, url_name, data=None, file_type=None, **kwargs):
+        retry = 0
+        url = parse.urljoin(base_url, url_name)
+        while True:
+            try:
+                await self._set_access_token()
```

Comment:
Can we call this method per session? Let's say we initiate it once before we start syncing. In case of expired token we catch the error and generate a new one.

Reply:
So, you mean we'll create an access token from `_get_session` method and when the token gets expired we'll create a new session to generate a new token. Is that what you want to do?

Reply:
What if we just don't try to reset the access token unless we first hit a 401 error saying the token is expired?

By doing this at the start of every API call, we're being pretty inefficient.

Reply:
If you look at the `_set_access_token`, we're not resetting the token each time we call a method instead we first checking if the existing access token still alive and if the token is expired, we're creating a new one else using the existing one.